esp-wifi/src/wifi/event.rs

```diff
@@ -0,0 +1,52 @@
+use core::cell::RefCell;
+
+use critical_section::Mutex;
+
+mod sealed {
+    pub trait Sealed {}
+}
+/// The type of handlers of events.
+pub type Handler<T> = dyn FnMut(&T) + Sync + Send;
+// fn default_handler<'a, T>(_: &'a T) {}
+/// Data for a wifi event which can be handled by an event handler.
+pub trait WifiEventData: sealed::Sealed + Sized + 'static {
+    /// Get the static reference to the handler for this event.
+    fn get_handler() -> &'static Mutex<RefCell<Option<&'static mut Handler<Self>>>>;
+
+    /// Get the handler for this event, replacing it with the default handler.
+    fn take_handler() -> Option<&'static mut Handler<Self>> {
+        critical_section::with(|cs| Self::get_handler().borrow_ref_mut(cs).take())
+    }
+    /// Set the handler for this event.
+    fn set_handler(f: &'static mut Handler<Self>) {
+        critical_section::with(|cs| Self::get_handler().borrow_ref_mut(cs).replace(f));
+    }
+    /// Atomic combination of `set_handler` and `get_handler`. Use this to add a
+    /// new handler while preserving the previous.
+    fn update_handler(
+        f: impl FnOnce(Option<&'static mut Handler<Self>>) -> &'static mut Handler<Self>,
+    ) {
+        critical_section::with(|cs| {
```

Comment:
Please just take `Box<dyn ...>` and store that in the static. Leaking the handlers seems error-prone as the caller must not drop the fake-static reference or they'll leak memory.

You may also want to pass the critical section to the callback.

Reply:
> fake-static

Why do you call it a fake-static? It's a real static. The memory has already leaked. The design involves leaking functions, since I don't know why one would want to drop the event handlers.

I didn't take a `Box` directly because I was going to remove the `Option` part of the type and instead give the user's function a nop `fn` first, so it is more uniform instead of checking if its `Some` in the user's handler. I was also going to add a `set_handler_fn` which would replace the callback with a provided `fn` after casting it to a `&'static dyn FnMut`, so the single-callback case doesn't need boxes.

Reply:
Also, why would you pass the critical section to the callback instead of letting the user use `critical_section::with` in the handler themselves? Is there a performance difference?

Reply:
Taking a critical section costs a few instructions even if nested, but passing around the ZST token should be free.